### PR TITLE
cnap/userv: add the implementation for websocket server

### LINKED ISSUE #61: design websocket server to subscribe result streams and publish to spa

### PR DESCRIPTION
**Pull request by Hulongyin**
This PR is to fix #61 , design and implement a websocket server to subscribe result streams and publish result streams to spa via websocket.

### CODE REVIEW

File: cnap/userv/websocket_server.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+        self._stream_broker_redis_host = self._get_env(
+            "STREAM_BROKER_REDIS_HOST", "127.0.0.1")
+        self._stream_broker_redis_port = self._get_env(
+            "STREAM_BROKER_REDIS_PORT", "6379")
+        self._pipelines = {}
+        self._users = {}
+
+    @staticmethod
+    def _get_env(key, default=None):
+        """
```
Comment (leyao-daily) [inline cnap/userv/websocket_server.py:45]: google docstring?
```
        Args:
            key (str): The name of the environment variable.
            default (str, optional): The default value to return if the environment variable does not exist.
        
        Returns:
            str: The value of the environment variable or the default value.
        """

File: cnap/userv/websocket_server.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        Connection handler for websocket server connection.
+
+        Args:
+            wsobj: the WebSocketServerProtocol object for websocket server connection.
+            path: the path for websocket server connection.
+        """
+        LOG.info("WebSocket server task start: %s path: %s", str(wsobj), path)
+        target = path[1:]  # skip / prefix
+
+        if len(list(self._users.keys())) > 100:
```
Comment (leyao-daily) [inline cnap/userv/websocket_server.py]: Make the number as a constant. `MAX_CLIENT_CONNECTIONS`

File: cnap/userv/websocket_server.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            LOG.error("Exceed the max number of client connections: 100.")
+            return
+
+        self._users[wsobj] = target
+        try:
+            while True:
+                try:
+                    _ = await wsobj.recv()
+                except asyncio.CancelledError:
```
Comment (leyao-daily) [inline cnap/userv/websocket_server.py:120]: I am thinking which is better combine the three to one to catch a broader range of exceptions or just leave it.

Comment (Hulongyin) [reply cnap/userv/websocket_server.py:120]: These three correspond to three different cases of exceptions, and will leave different logs to make it easier to locate errors?

File: cnap/userv/websocket_server.py (hunk 1 above)
Comment (Hulongyin) [reply cnap/userv/websocket_server.py:45]: Thanks, fixed.

File: cnap/userv/websocket_server.py (hunk 2 above)
Comment (Hulongyin) [reply cnap/userv/websocket_server.py]: Thanks, fixed.

**Review by leyao-daily — APPROVED** [review]
LGTM.

**Pull request merged**